DistroLauncher/ApplicationStrategy.h

```diff
@@ -0,0 +1,113 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#pragma once
+
+#include "state_machine.h"
+#include "installer_policy.h"
+#include "installer_controller.h"
+
+#ifndef _M_ARM64
+#include <mutex>
+#include "local_named_pipe.h"
+#include "console_service.h"
+#include "splash_controller.h"
+
+namespace Oobe
+{
+    // This strategy fulfills the essential API required by the Oobe::Application<> class and augment it with the
+    // machinery necessary to coordinate the splash screen application, the console service for console redirection and
+    // visibility toggling and the OOBE itself running inside Ubuntu.
+    class x64Strategy
+    {
+      private:
+        // This mutex must be acquired before calling ConsoleService::redirectConsole() and
+        // ConsoleService::restoreConsole(). That's a consequence of using RegisterWaitSingleObject to notify if the
+        // user closes the splash application. That function imposes concurrent access to this class and could cause
+        // races and deadlocks if all stars and planets aligned.
+        std::timed_mutex consoleGuard;
+        bool consoleIsVisible = true;
+        bool splashIsRunning = false;
+        std::filesystem::path splashExePath;
+
+        HANDLE consoleReadHandle = nullptr;
+        InstallerController<> installer;
+        // optionals for lazy creation.
+        std::optional<SplashController<>> splash;
+        std::optional<Win32Utils::ConsoleService<Win32Utils::LocalNamedPipe>> console;
+
+        /// Restores the console context, which means undo the console redirection and making the window visible.
+        void do_show_console();
+
+        /// Hides the splash window.
+        void do_toggle_splash();
+
+        /// Notifies the splash controller to quit the application. Console restoration is immediately requested to
+        /// prevent leaving the launcher process without a stdout consumer.
+        void do_close_splash();
+
+      public:
+        /// Places the sequence of events to make the OOBE perform an automatic installation.
+        HRESULT do_autoinstall(std::filesystem::path autoinstall_file);
+
+        /// Places the sequence of events to make the OOBE perform an interactive installation.
+        HRESULT do_install();
+
+        /// Places the sequence of events to make the OOBE perform an automatic installation.
+        HRESULT do_reconfigure();
+
+        /// Triggers the console redirection and launch the splash screen application.
+        void do_run_splash();
+
+        x64Strategy(Win32Utils::LocalNamedPipe&& pipe);
```

Comment:
Shouldn’t we name that something else than x64… Imagine that tomorrow flutter is available on arm64, or even better, that a new arch appears in the wild and that we can run flutter on it!

Probably something like splashStrategy and noSplashStrategy will make it more future-proof?

Reply:
I Agree. I made them `SplashEnabledStrategy` and `NoSplashStrategy`.